City-of-Helsinki/drupal-hdbt
# UHF-7115 search accessibility

# [UHF-7115](https://helsinkisolutionoffice.atlassian.net/browse/UHF-7115)

* Search form was missing an `role` attribute for accessibility
* When menu API is misbehaving, the console message was unhelpful
* When menu API was misbehaving and javascript toggled fallback mode on mobile and mega menus, user was unable to close either.
* Search form focusing on menu open got broken on a refactoring of code
* On mobile landscape screens, some of the dropdowns were unscrollable due to overmatching selectors
* On breakpoint L, language menu disappeared when mobile menu was opened even tho language menu was not visible in menu.

## What was done

* Role attribute was added to hero and header search forms
* API error message was improved
* When menu API was misbehaving and javascript toggled fallback mode on mobile and mega menus, user can now close both mobile and megamenu
* Search form focusing on menu open was fixed
* Dropdowns should now be scrollable when needed both in js and non-js modes
* On breakpoint L, language menu no longer erroneously disappears on menu open.

## How to install

* Make sure your instance is up and running on latest dev branch.
  * `git pull origin dev`
  * `make fresh`
* Update the HDBT theme
  * `composer require drupal/hdbt:dev-UHF-7115_search_accessibility`
* Run `make drush-cr`

## How to test


* [ ] Check that header and hero forms contain `role="search"` attribute
* [ ] Check the API error message in devtools console.
    * For example on chrome browser you can block menu API call with `Network request blocking` that can be found with the `ESC`-key toggleable console, from the `⋮`-button
    * Open `Network request blocking`-view
![image](https://user-images.githubusercontent.com/1191667/230084540-567b9c73-934a-4474-8e26-2e02e07e4413.png)
    * Add `https://helfi-etusivu.docker.so/fi/api/v1/global-mobile-menu` to block list
![image](https://user-images.githubusercontent.com/1191667/230084779-c160ac79-3728-4654-ac5f-34fb6ab2d7a3.png)
    * Reload page, check error message from console tab. Remember to disable `Network request blocking`
* [ ] Check that mega and mobile menus in API fallback mode can be opened and closed properly. (See details for blocking API in previous step)
* [ ] Check that search input is focused when search dropdown is opened
* [ ] On mobile landscape mode, check that all three dropdown menus can be scrolled and the underlying body does not scroll in the background at the same time.
* [ ] Check that on the last breakpoint where mobile menu is visible, the language menu does not disappear when mobile menu is opened.

## Designers review




[UHF-7115]: https://helsinkisolutionoffice.atlassian.net/browse/UHF-7115?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
